source/MaterialXView/ShaderValidators/Glsl/GlslValidator.h

```diff
@@ -0,0 +1,326 @@
+
+    /// This method can be used to incrementally set the code stages before validation.
+    /// @param code Shader code string for a given stage
+    /// @param stage Shader stage. See stages allowed in the class HwShader.
+    void setStage(const std::string& code, size_t stage);
+
+    /// Get code string for a given stage
+    /// @return Shader stage string. String is empty if not found.
+    const std::string getStage(size_t stage) const;
+
+    /// Get the number of stages
+    /// @return Stage count
+    size_t numberofStages() const
+    {
+        return HwShader::NUM_STAGES;
+    }
+
+    /// Clear out any existing stages
+    void clearStages();
+
+    /// @}
+    /// @name Program handling
+    /// @{
+
+    /// Create the shader program from stages specified
+    /// An exception is thrown if the program cannot be created.
+    /// The exception will contain a list of program creation errors.
+    /// @return Program identifier. 
+    unsigned int createProgram();
+
+    /// Get list of program input uniforms. 
+    /// The program must have been created successfully first.
+    /// An exception is thrown if the parsing of the program for uniforms cannot be performed.
+    /// @return Program uniforms list.
+    const ProgramInputMap& getUniformsList();
+
+    /// Get list of program input attributes. 
+    /// The program must have been created successfully first.
+    /// An exception is thrown if the parsing of the program for attribute cannot be performed.
+    /// @return Program attributes list.
+    const ProgramInputMap& getAttributesList();
+
+    /// Delete any currently created shader program
+    void deleteProgram();
+
+    /// @}
+    /// @name Visualization
+    /// @{
+
+    /// Render to buffer
+    void render();
+
+    /// Save buffer to disk
+    /// @param fileName Name of file to save rendered image to.
+    /// @param imageHandler Handler used to save image
+    /// @return true if successful
+    void save(std::string& fileName, const ImageHandlerPtr imageHandler);
+
+    /// @}
+
+  protected:
+    /// Internal cleanup of stages and OpenGL constructs
+    void cleanup();
+
+    /// Check if there is a valid set of stages to build program from
+    bool haveValidStages() const;
+
+    /// @name Target handling
+    /// @{
+
+    /// Create a offscreen target used for rendering.
+    bool createTarget();
+    /// Delete any created offscreen target.
+    void deleteTarget();
+    /// Bind or unbind any created offscree target.
+    bool bindTarget(bool bind);
+
+    /// @}
+    /// @name Program introspection
+    /// @{
+
+    /// Update a list of program input uniforms
+    const ProgramInputMap& updateUniformsList();
+
+    /// Update a list of program input attributes
+    const ProgramInputMap& updateAttributesList();
+
+    /// Clear out any cached input lists
+    void clearInputLists();
+
+    /// Index used to access cached attribute buffers
+    enum AttributeIndex {
```

Comment:
This is to keep a mapping between program attributes and actual h/w buffers to bind. This is basically using the types we talked about where there is no additional data massaging required to route to h/w inputs with either greather or less channels.